### DISCUSSION TITLE
✅ Product Roadmap (as of March 2026)

### DISCUSSION THREAD

**Discussion by yorek** [Azure/data-api-builder#1377] [Announcements]
Here's the planned roadmap. We'll update this post as soon as we move forward with the development. Subject to change.

## Release 2.3 (December 2026)

https://github.com/Azure/data-api-builder/issues?q=is%3Aissue%20state%3Aopen%20label%3A2.3

* MCP Server Tool: Persisted Documents #2840
* MCP Server Tool: JOIN support / GraphQL query #3208

### Under investigation:

https://github.com/Azure/data-api-builder/issues?q=is%3Aissue%20state%3Aopen%20label%3A2.4

* Add `Geospatial` data type support #3219
* Add `Geometry` data type support #3220
* Add `rowversion` data type support #3222
* Add `sql_variant` data type support #3221
* Add REST keyword `$count` #3218
* Configuration function: `@sql()` #3224
* Auto-config (auto relationships) #3207
* Auto-config (views) #3225 
* Auto-config (stored procedures) #3224 
* CLI enhancements #3191
* Support for Auth 2.1 #3227 

## Release 2.2 (October 2026)

[https://github.com/Azure/data-api-builder/issues?q=is%3Aissue%20state%3Aopen%20label%3A2.2](https://github.com/Azure/data-api-builder/issues?q=is%3Aissue%20state%3Aopen%20label%3A2.2)

* Pagination / Pagination metadata #2459 
* Pagination / New pagination keywords: #3198
* Relationships between views #3079
* Keyless views (read-only) #2791 
* DAB validate CI/CD #2923 
* Health UI #3223

## Release 2.1 (August 2026)

[https://github.com/Azure/data-api-builder/issues?q=is%3Aissue%20state%3Aopen%20label%3A2.1](https://github.com/Azure/data-api-builder/issues?q=is%3Aissue%20state%3Aopen%20label%3A2.1)

* Optionally omit unsupported columns #3181
* Add `Vector` data type support #2767
* Add `JSON` data type support #2768
* Add `XML` data type support #2769
* Telemetry in CLI #3214
* Telemetry in engine #3215
* Telemetry in application_name #3216

### External contribution

* Embedding subsystem #3104
* Embedding / Semantic caching #3206
* Embedding / Parameter substitution #3204

## Release 2.0 (June 2026)

<img width="285" height="284" alt="image" src="https://github.com/user-attachments/assets/2ef6b4f2-7a77-4cf4-b198-d42d9d7bd53a" />

* ✅ MCP Server Tool: Stored Procedures #2864
* ✅ MCP Server Tool: `aggregate_records` #3199
* ✅ OBO/pass-through authentication #2898
* ✅ Auth provider: `Unauthenticated` #3075
* ✅ Auth / Roll inheritance #3164
* ✅ Auto-config (auto entities) #3127
* ✅ HTTP Response Compression #3003
* ✅ REST Advanced Paths #2999
* ✅ ⭐ Developer Default Image #3168
* ✅ Role-based OpenAPI #2998
* ✅ Enhanced OpenTelemetry #3111
* ✅ Enhanced Custom Logs
* ✅ Enhanced REST filters #3080
* ✅ Enhanced PATCH support #3150

## Release 1.7 (March 2026)

* ✅ MCP Server Streamable HTTP #2866
* ✅ MCP Server STDIO #3050
* ✅ MCP Server Tool: `describe_entities` #3043
* ✅ MCP Server Tool: `create_record` #2828
* ✅ MCP Server Tool: `read_records` #2833
* ✅ MCP Server Tool: `update_record` #2829
* ✅ MCP Server Tool: `delete_record` #2830
* ✅ MCP Server Tool: `execute_entity` #2831
* ✅ Server semantic description #2937
* ✅ Entity semantic description #2871
* ✅ Field semantic description #2836
* ✅ Parameter semantic description #2904
* ✅ Support Azure Key Vault (@akv()) #2708
* ✅ New Entity `Parameters` [array] #2365
* ✅ New Entity `Fields` [array]

## Release 1.6 (July 2025)

* ✅ Adding File Log Sink
* ✅ Log level custom configuration
* ✅ Level 2 Cache (Redis)
* ✅ Expand `dab configure` support
* ✅ Azure Log Analytics support

## Release 1.5 (May 2025)

* ✅ GraphQL Aggregations (👷‍♂️: Fabric)
* ✅ Authentication Synonyms (👷‍♂️: Community)
* ✅ Custom OTEL Traces in REST (👷‍♂️: Pasta)
* ✅ Custom OTEL Traces in GraphQL

### Other improvements

* ✅ Health Endpoints
* ✅ Level 2 Cache (👷‍♂️: Pasta)
* ✅ Caching Header Support

## Release 1.4 (Jan 2025)

* ✅ Open Telemetry Support
* ✅ CLI `dab configure` support
* ✅ .NET POCO support with strict bodies
* ✅ Prototype VS Code Extensions

## 1.3 (September 2024)

[https://github.com/Azure/data-api-builder/releases/tag/v1.3.19](https://github.com/Azure/data-api-builder/releases/tag/v1.3.19)

* ✅ Hot Reload of the Configuration file (AB) [https://github.com/Azure/data-api-builder/issues?q=is%3Aissue+is%3Aopen+reload](https://github.com/Azure/data-api-builder/issues?q=is%3Aissue+is%3Aopen+reload)
* ✅ Include `runtime` support
* ✅ Include `entity` support (tentative)
* ✅ Introduce `dab configure` to the CLI (AK) #2251
* ✅ Expand Cache implementation (SL) #2253
* ✅ HTTP request header `no-cache`, `no-store`
* ✅ Add CLI implementation for cache in `runtime` and `entity`
* ✅ Default runtime cache `enabled` to true when enabled an `entity` (in CLI)
* ✅ Add support for `stored-procedure` entities

## 1.2 (July 2024) Stabilization release

* ✅ Pre-release v1.2.9-rc
  [https://github.com/Azure/data-api-builder/releases/tag/v1.2.9-rc](https://github.com/Azure/data-api-builder/releases/tag/v1.2.9-rc)

## 1.1 (May 2024)

* ✅ Complete multiple create support for GQL (A&S)
* ✅ Upgrade to .NET 8 (SL)

## 0.11 (January 2024)

* ✅ Support multi-data sources and config files (RK)
* ✅ Support for Synapse (RK)

## 0.10 (November 2023)

* ✅ Caching support for REST and GQL endpoints (SL)
* ✅ Partial nested mutation support for GQL (A&S)
* ✅ Improved support for `time` data type in `mssql` (AA)
* ✅ Support `validate` in the CLI (AK)

## 0.9 (September 2023)

* ✅ Add support for Application Insights
* ✅ Mutations on table with triggers
* ✅ Support for ignoring extraneous fields in REST request body
* ✅ Adding support for positive `boolean` options in CLI
* ✅ Structured logging compatibility
* ✅ Adding Application Name for `mssql` connections
* ✅ Support `time` data type in `mssql`

## 0.8 (July 2023)

* ✅ Add support for `.env` file

## 0.7 (May 2023)

* ✅ Support for OpenAPI v3
* ✅ Executing GraphQL and REST mutations in a transaction
* ✅ Add support for Cosmos field level authorization
* ✅ Database policy support for PUT/PATCH operations
* ✅ Adding `Merge` config capabilities to CLI

## Top of backlog

* Custom hook at post-db query 
* Custom scripting at pre-db query 
* Custom error messages
* Output parameters
* Add `regex` operator: `$filter=col regex pattern`
* Custom scripting at post query (Fab)
* Substring operations in item-level security
* Configuration UI / editor
* Create dab-config.json from dacpac
* Enable caching with Session Context
* Additional native authentication providers
* Token claims mocking for testing loop
* Native integration with Azure Container Apps
* Native integration with Azure Functions
* Subscriptions for GraphQL endpoints
* REST support for new `$value` scalar-value keyword
* Introduce OData endpoint

**Comment by zebleck** [comment]
Would be great if policies allowed for checking many-to-many relationships. For example, we want to only allow reading an entity if its many-to-many relationship with the User entity contains a user with @claims.userId.
